actionpack/test/dispatch/routing/direct_url_helpers_test.rb

```diff
@@ -0,0 +1,215 @@
+require "abstract_unit"
+
+class TestDirectUrlHelpers < ActionDispatch::IntegrationTest
+  class Linkable
+    attr_reader :id
+
+    def initialize(id)
+      @id = id
+    end
+
+    def linkable_type
+      self.class.name.demodulize.underscore
+    end
+  end
+
+  class Category < Linkable; end
+  class Collection < Linkable; end
+  class Product < Linkable; end
+
+  class Model
+    extend ActiveModel::Naming
+    include ActiveModel::Conversion
+
+    attr_reader :id
+
+    def initialize(id = nil)
+      @id = id
+    end
+
+    def model_name
+      @_model_name ||= ActiveModel::Name.new(self.class, nil, self.class.name.demodulize)
+    end
+
+    def persisted?
+      false
+    end
+  end
+
+  class Basket < Model; end
+  class User < Model; end
+  class Video < Model; end
+
+  Routes = ActionDispatch::Routing::RouteSet.new
+  Routes.draw do
+    default_url_options host: "www.example.com"
+
+    root to: "pages#index"
+    get "/basket", to: "basket#show", as: :basket
+    get "/profile", to: "users#profile", as: :profile
+    get "/media/:id", to: "media#show", as: :media
+
+    resources :categories, :collections, :products
+
+    namespace :admin do
+      get "/dashboard", to: "dashboard#index"
+    end
+
+    direct(:website)  { "http://www.rubyonrails.org" }
+    direct("string")  { "http://www.rubyonrails.org" }
+    direct(:helper)   { basket_url }
+    direct(:linkable) { |linkable| [:"#{linkable.linkable_type}", { id: linkable.id }] }
+    direct(:params)   { |params| params }
+    direct(:symbol)   { :basket }
+    direct(:hash)     { { controller: "basket", action: "show" } }
+    direct(:array)    { [:admin, :dashboard] }
+    direct(:options)  { |options| [:products, options] }
+    direct(:defaults, size: 10) { |options| [:products, options] }
+
+    direct(class: "Basket") { |basket| [:basket] }
+    direct(class: "User", anchor: "details") { |user, options| [:profile, options] }
+    direct(class: "Video") { |video| [:media, { id: video.id }] }
```

Comment:
Ah... ouch.

So there's no way to really delegate to another route helper directly.

Could we make `media_path(video)` work, *and* have that magically behave like media_url or media_path depending on whether we're doing video_url or video_path, a bit like ActionMailer already does by forcing a full URL even when you ask for just the path?